src/algebra/order/hom/basic.lean

```diff
@@ -0,0 +1,60 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import algebra.hom.group
+import algebra.order.with_zero
+import order.hom.basic
+
+/-!
+# Algebraic order homomorphism classes
+
+This file defines hom classes for common properties at the intersection of order theory and algebra.
+
+## Typeclasses
+
+* `nonneg_hom_class`: Homs are nonnegative: `∀ f a, 0 ≤ f a`
+* `subadditive_hom_class`: Homs are subadditive: `∀ f a b, f (a + b) ≤ f a + f b`
+* `submultiplicative_hom_class`: Homs are submultiplicative: `∀ f a b, f (a * b) ≤ f a * f b`
+* `mul_le_add_hom_class`: `∀ f a b, f (a * b) ≤ f a + f b`
+-/
+
+set_option old_structure_cmd true
+
+open function
+
+variables {F α β γ δ : Type*}
+
+/-- `nonneg_hom_class F α β` states that `F` is a type of nonnegative morphisms. -/
+class nonneg_hom_class (F : Type*) (α β : out_param $ Type*) [has_zero β] [has_le β]
+  extends fun_like F α (λ _, β) :=
+(map_nonneg (f : F) : ∀ a, 0 ≤ f a)
+
+/-- `subadditive_hom_class F α β` states that `F` is a type of subadditive morphisms. -/
+class subadditive_hom_class (F : Type*) (α β : out_param $ Type*) [has_add α] [has_add β] [has_le β]
+  extends fun_like F α (λ _, β) :=
+(map_add_le_add (f : F) : ∀ a b, f (a + b) ≤ f a + f b)
```

Comment:
This is just bikeshedding, but I think it's reasonable to shorten this to `map_add_le`.
```suggestion
(map_add_le (f : F) : ∀ a b, f (a + b) ≤ f a + f b)
```

Reply:
I kept them long to avoid ambiguity with `map_mul_le_add` (and the eventual `map_add_le_max` and such).